src/extensions/surveys.tsx

```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
+    posthog.capture('survey dismissed', {
+        $survey_name: survey.name,
+        $survey_id: survey.id,
+        sessionRecordingUrl: posthog.get_session_replay_url?.(),
+        $set: {
+            [`$survey_dismissed/${survey.id}`]: true,
+        },
+    })
+    localStorage.setItem(`seenSurvey_${survey.id}`, 'true')
+    window.dispatchEvent(new Event('PHSurveyClosed'))
+}
+
+export function OpenTextQuestion({
+    question,
+    appearance,
+    onSubmit,
+    closeSurveyPopup,
+}: {
+    question: BasicSurveyQuestion
+    appearance: SurveyAppearance
+    onSubmit: (text: string) => void
+    closeSurveyPopup: () => void
+}) {
+    const textRef = Preact.createRef()
+    const [text, setText] = useState('')
+    return (
+        <div
+            className="survey-box"
+            style={{ backgroundColor: appearance.backgroundColor || defaultBackgroundColor }}
+            ref={textRef}
+        >
+            <Cancel onClick={() => closeSurveyPopup()} />
+            <QuestionHeader
+                question={question.question}
+                description={question.description}
+                backgroundColor={appearance.backgroundColor}
+            />
+            <textarea rows={4} placeholder={appearance?.placeholder} onInput={(e) => setText(e.currentTarget.value)} />
+            <BottomSection
+                text={question.buttonText || 'Submit'}
+                submitDisabled={!text && !question.optional}
+                appearance={appearance}
+                onSubmit={() => onSubmit(text)}
+            />
+        </div>
+    )
+}
+
+export function LinkQuestion({
+    question,
+    appearance,
+    onSubmit,
+    closeSurveyPopup,
+}: {
+    question: LinkSurveyQuestion
+    appearance: SurveyAppearance
+    onSubmit: (clicked: string) => void
+    closeSurveyPopup: () => void
+}) {
+    return (
+        <div className="survey-box">
+            <Cancel onClick={() => closeSurveyPopup()} />
+            <QuestionHeader question={question.question} description={question.description} />
+            <BottomSection
+                text={question.buttonText || 'Submit'}
+                submitDisabled={false}
+                link={question.link}
+                appearance={appearance}
+                onSubmit={() => onSubmit('link clicked')}
+            />
+        </div>
+    )
+}
+
+export function RatingQuestion({
+    question,
+    questionIndex,
+    appearance,
+    onSubmit,
+    closeSurveyPopup,
+}: {
+    question: RatingSurveyQuestion
+    questionIndex: number
+    appearance: SurveyAppearance
+    onSubmit: (rating: number | null) => void
+    closeSurveyPopup: () => void
+}) {
+    const threeScaleEmojis = [dissatisfiedEmoji, neutralEmoji, dissatisfiedEmoji]
+    const fiveScaleEmojis = [veryDissatisfiedEmoji, dissatisfiedEmoji, neutralEmoji, satisfiedEmoji, verySatisfiedEmoji]
+    const fiveScaleNumbers = [1, 2, 3, 4, 5]
+    const tenScaleNumbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

Comment:
These are all constants that should be moved to a constants.ts file or something. It's not the biggest deal in the world that these are declared in each function call but it's far from best practice.

Reply:
I wasn't able to pull these into a constants file because of the Preact separate components issue as commented above, but I did take them out of the function call.